web/pandas/pdeps/0005-no-default-index-mode.md

```diff
@@ -0,0 +1,189 @@
+
+    ```python
+    In [42]: pd.concat([ser1, ser2], axis=1, ignore_index=True)
+    Out[42]:
+          0     1
+    1  10.0  10.0
+    2  15.0  15.0
+    3  20.0  20.0
+    5  25.0   NaN
+    4   NaN  25.0
+    ```
+
+- it can be frustrating to have to repeatedly call `.reset_index()` (https://twitter.com/chowthedog/status/1559946277315641345):
+
+    ```python
+    In [45]: df.value_counts(['sender', 'receiver']).reset_index().rename(columns={0: 'count'})
+    Out[45]:
+       sender  receiver  count
+    0       1         1      1
+    1       2         2      1
+    2       3         3      1
+    3       5         4      1
+    ```
+
+With this option enabled, users who don't want to worry about indices wouldn't need to.
+
+## Detailed Description
+
+This would require 3 steps:
+1. creation of a ``NoIndex`` object, which would be a subclass of ``RangeIndex`` on which
+  some operations such as ``append`` would behave differently.
+  The ``default_index`` function would then return ``NoIndex`` (rather than ``RangeIndex``) if this mode is enabled;
+2. adjusting ``DataFrameFormatter`` and ``SeriesFormatter`` to not print row labels for objects with a ``NoIndex``;
+3. adjusting methods which currently return an index to just insert a new column instead.
+
+Let's expand on all three below.
+
+### 1. NoIndex object
+
+Most of the logic could be handled within the ``NoIndex`` object.
+It would be like a ``RangeIndex``, but with the following differences:
+- `name` could only be `None`;
+- `start` could only be `0`, `step` `1`;
+- when appending an extra element, the new `Index` would still be `NoIndex`;
+- when slicing, one would still get a `NoIndex`;
+- two ``NoIndex`` objects can't be aligned. Either they're the same length, or pandas raises;
+- aligning a ``NoIndex`` object with one which has an index will raise, always;
+- ``DataFrame`` columns can't be `NoIndex` (so ``transpose`` would need some adjustments when called on a ``NoIndex`` ``DataFrame``);
+- `insert` and `delete` should raise. As a consequence, `.drop` with `axis=0` would always raise;
+- arithmetic operations (e.g. `NoIndex(3) + 2`) would all raise.
+
+### 2. DataFrameFormatter and SeriesFormatter changes
+
+When printing an object with a ``NoIndex``, then the row labels wouldn't be shown:
+
+```python
+In [14]: pd.set_option('mode.no_default_index', True)
+
+In [15]: df = pd.DataFrame({'a': [1,  2, 3], 'b': [4, 5, 6], 'c': [7, 8, 9]})
+
+In [16]: df
+Out[16]:
+ a  b  c
+ 1  4  7
+ 2  5  8
+ 3  6  9
+```
+
+### 3. Nobody should get an index unless they ask for one
+
+The following would work in the same way:
+```python
+pivot = (
+    pd.pivot_table(df, values="D", index=["A", "B"], columns=["C"], aggfunc=np.sum)
+).reset_index()
+
+with pd.option_context('mode.no_default_index', True):
+    pivot = (
+        pd.pivot_table(df, values="D", index=["A", "B"], columns=["C"], aggfunc=np.sum)
+    )
+```
```

Comment:
Wouldn't the first example create a `RangeIndex` due to calling `reset_index()` ?  With the second example, you're explicitly creating the index, so why with `no_default_index` cause that not to happen?  I'm not following your logic here.

I think you also have to say that the behavior of `reset_index()` would change.  Without that option, we create a default `RangeIndex`.  With `no_default_index`, you would create a `NoIndex`.  So this behavior change should appear in the PDEP

Reply:
Under this mode, the `default_index` function would return a `NoIndex`. `reset_index()` calls `default_index`, and so `reset_index()` would return a `NoIndex`

I'll clarify, thanks for your review!